presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -373,28 +383,88 @@ private void loadPartition(HivePartitionMetadata partition)
+        // Sort FileStatus objects (instead of, e.g., fileStatus.getPath().toString). This matches org.apache.hadoop.hive.ql.metadata.Table.getSortedPaths
+        fileStatuses.sort(null);
+
+        List<Set<LocatedFileStatus>> list = null;
+        if (fileStatuses.size() == bucketCount) {
+            list = getBucketFilesSetsSingle(fileStatuses);
+        }
+        else if (isMultiFileBucketingEnabled(session)) {
+            list = getBucketFilesSetsMulti(fileStatuses);
+            if (list.size() != bucketCount) {
+                list = null;
+            }
+        }
+
+        if (list == null) {
+            throw new PrestoException(HIVE_INVALID_BUCKET_FILES,
+                    format("%s The number of files in the directory (%s) does not match the declared bucket count (%s) for partition: %s",
+                            CORRUPT_BUCKETING,
+                            fileStatuses.size(),
+                            bucketCount,
+                            hiveFileIterator.getPartitionName()));
         }
 
-        // Sort FileStatus objects (instead of, e.g., fileStatus.getPath().toString). This matches org.apache.hadoop.hive.ql.metadata.Table.getSortedPaths
-        list.sort(null);
         return list;
     }
 
+    private List<Set<LocatedFileStatus>> getBucketFilesSetsSingle(List<LocatedFileStatus> sortedFileStatuses)
+    {
+        return sortedFileStatuses.stream()
+                .map(ImmutableSet::of)
+                .collect(Collectors.toList());
+    }
+
+    private List<Set<LocatedFileStatus>> getBucketFilesSetsMulti(List<LocatedFileStatus> sortedFileStatuses)
+    {
+        ImmutableList.Builder<Set<LocatedFileStatus>> list = ImmutableList.builder();
+        ImmutableSet.Builder<LocatedFileStatus> bucketSet = ImmutableSet.builder();
+        String lastBucketId = null;
+
+        for (LocatedFileStatus next : sortedFileStatuses) {
+            String currentBucketId = getBucketId(next.getPath().getName());
+            if (lastBucketId != null && !lastBucketId.equals(currentBucketId)) {
+                list.add(bucketSet.build());
+                bucketSet = ImmutableSet.builder();
+            }
+            bucketSet.add(next);
+            lastBucketId = currentBucketId;
+        }
+        if (lastBucketId != null) {
+            list.add(bucketSet.build());
+        }
+        return list.build();
+    }
+
+    private String getBucketId(String fileName)
+    {
+        // this matches the observed Hive's behaviour during insert. Then files are named using M/R fremwork using
+        // XXXXXXXX_Y_SUFFIX pattern where:
+        // - XXXXXXXX - is task id
+        // - Y - is attempt id
+        // - SUFFIX - is optional copy_N suffix
+
+        // this method will only be used if number of files in directory does not match number of partitions
+        // even if multi file bucketing is enabled.
+
+        String bucketId = Iterables.getFirst(BUCKET_ID_SPLITTER.split(fileName), "");
```

Comment:
I would parse that using regexp and throw an exception if it doesn't match to improve our fail fast capabilities.